scripts/train_cz_gate.py

```diff
@@ -0,0 +1,273 @@
+LR = 0.00025
+ITERATIONS = 1000
+ALPHA = 10
+BETA = 10_000
+S_STAR = 0.075
+
+tf.get_logger().setLevel("ERROR")
+
+
+def get_expected_density_matrix(
+    state_vector: tf.Tensor, cutoff: int, calculator: pq.TensorflowCalculator
+) -> tf.Tensor:
+    state_00 = [0, 1, 0, 1]
+    state_01 = [0, 1, 1, 0]
+    state_10 = [1, 0, 0, 1]
+    state_11 = [1, 0, 1, 0]
+
+    config = pq.Config(normalize=False, cutoff=cutoff)
+    expected_program = pq.Program(
+        instructions=[
+            pq.StateVector(state_00, coefficient=state_vector[0]),
+            pq.StateVector(state_01, coefficient=state_vector[1]),
+            pq.StateVector(state_10, coefficient=state_vector[2]),
+            pq.StateVector(state_11, coefficient=-state_vector[3]),
+        ]
+    )
+
+    simulator = pq.PureFockSimulator(d=4, config=config, calculator=calculator)
+    expected_state = simulator.execute(expected_program).state
+
+    return expected_state.density_matrix
+
+
+def get_ideal_weights():
+    calculator = pq.NumpyCalculator()
+    modes = (0, 2, 4, 5)
+
+    U = np.array(
+        [
+            [-1 / 3, -np.sqrt(2) / 3, np.sqrt(2) / 3, 2 / 3],
+            [np.sqrt(2) / 3, -1 / 3, -2 / 3, np.sqrt(2) / 3],
+            [
+                -np.sqrt(3 + np.sqrt(6)) / 3,
+                np.sqrt(3 - np.sqrt(6)) / 3,
+                -np.sqrt((3 + np.sqrt(6)) / 2) / 3,
+                np.sqrt(1 / 6 - 1 / (3 * np.sqrt(6))),
+            ],
+            [
+                -np.sqrt(3 - np.sqrt(6)) / 3,
+                -np.sqrt(3 + np.sqrt(6)) / 3,
+                -np.sqrt(1 / 6 - 1 / (3 * np.sqrt(6))),
+                -np.sqrt((3 + np.sqrt(6)) / 2) / 3,
+            ],
+        ]
+    )
+
+    V = calculator.embed_in_identity(U, get_operator_index(modes), 6)
+
+    return get_weights_from_interferometer(V, calculator)
+
+
+def _calculate_loss(
+    weights: tf.Tensor,
+    expected_density_matrix: tf.Tensor,
+    P: tf.Tensor,
+    calculator: pq.TensorflowCalculator,
+    state_vector: tf.Tensor,
+    cutoff: int,
+    alpha: tf.Tensor,
+    beta: tf.Tensor,
+    S_star: tf.Tensor,
+):
+    d = 6
+    config = pq.Config(cutoff=cutoff, normalize=False)
+    np = calculator.np
+
+    ancilla_modes = (4, 5)
+
+    state_00 = [0, 1, 0, 1]
+    state_01 = [0, 1, 1, 0]
+    state_10 = [1, 0, 0, 1]
+    state_11 = [1, 0, 1, 0]
+
+    ancilla_state = [1, 1]
+
+    preparation = pq.Program(
+        instructions=[
+            pq.StateVector(state_00 + ancilla_state, coefficient=state_vector[0]),
+            pq.StateVector(state_01 + ancilla_state, coefficient=state_vector[1]),
+            pq.StateVector(state_10 + ancilla_state, coefficient=state_vector[2]),
+            pq.StateVector(state_11 + ancilla_state, coefficient=state_vector[3]),
+        ]
+    )
+
+    interferometer = get_interferometer_from_weights(
+        weights, d, calculator, config.complex_dtype
+    )
+
+    program = pq.Program(
+        instructions=[
+            *preparation.instructions,
```

Comment:
I think it is not necessary to create `preparation` here. You can just paste the `pq.StateVector` instructions here.